lib/coap/transport.ex

```diff
@@ -0,0 +1,192 @@
+defmodule CoAP.Transport do
+  @moduledoc """
+  Implements CoAP message layer
+  """
+  use GenStateMachine
+
+  alias CoAP.Message
+
+  @ack_timeout 2_000
+  @ack_random_factor 1.5
+  @max_retransmit 4
+
+  defstruct socket: nil,
+            socket_ref: nil,
+            peer: nil,
+            retries: 0,
+            max_retransmit: @max_retransmit,
+            retransmit_timeout: 0,
+            ack_timeout: @ack_timeout,
+            ack_random_factor: @ack_random_factor,
+            max_transmit_span: 0,
+            transport_opts: nil
+
+  @type t :: %__MODULE__{
+          socket: pid() | nil,
+          socket_ref: reference() | nil,
+          peer: %URI{} | nil,
+          retries: integer(),
+          max_retransmit: integer(),
+          retransmit_timeout: integer(),
+          ack_timeout: integer(),
+          ack_random_factor: integer() | float(),
+          max_transmit_span: integer(),
+          transport_opts: any()
+        }
+
+  @type host :: String.Chars.t() | :inet.ip_address()
+  @type arg() ::
+          {:peer, {host(), integer()}}
+          | {:transport_opts, any()}
+          | {:max_retransmit, integer()}
+          | {:ack_timeout, integer()}
+          | {:ack_random_factor, integer() | float()}
+  @type args() :: arg()
+
+  @type peer() :: URI.t()
+  @type transport() :: pid()
+  @type transport_opts() :: any()
+
+  # Socket implementation is not linked to transport process, but monitored
+  @callback start({peer(), transport(), transport_opts()}) :: GenServer.on_start()
+
+  @doc false
+  @spec start_link(args()) :: GenStateMachine.on_start()
+  def start_link(args) do
+    GenStateMachine.start_link(__MODULE__, args)
+  end
+
+  @impl GenStateMachine
+  def init(args) do
+    args = Enum.into(args, %{})
+
+    with args <- Enum.into(args, %{}),
```

Comment:
Never been a fan of `with` or the `{:ok, _}` everywhere pattern. Much prefer it as a pipeline that takes args, each function adding to the state, and then returning state at the end. Kind of an accumulator pattern.

Reply:
I was hesitating but I prefer accumulator pattern me too :)